active/003-seq-feature.md

```diff
@@ -0,0 +1,121 @@
+For example:
+
+```python
+from pandas.core.interval import Interval, IntervalIndex
+intervals = [(190, 255),
+             (190, 255),
+             (250, 2799),
+             (337, 2799),
+             (2801, 3733)]
+iid = IntervalIndex.from_tuples(intervals)
+df = pd.DataFrame(data={'feature': ['gene1', 'gene2', 'gene3', 'gene4', 'gene5']},
+                  index=iid)
+df.loc[254, :]
+```
+Because its implementation of `IntervalTree`, it will return all the features containing
+coordinate of 254 very fast:
+```
+            feature
+(190, 255]    gene1
+(190, 255]    gene2
+(250, 2799]   gene3
+```
+
+# Detailed design
+
+Use `IntervalIndex` in `pandas`.  The IntervalIndex is functional for our purposes, and we don't have to even maintain it.
+
+The `positional_metadata` dataframe has an IntervalIndex as its row index and has feature metadata as column(s) so that each row is sequence region and a feature can have multiple rows/regions. The `positional_metadata` can have duplicate row index to accommandate the scenario that a single region belongs to multiple features.
+
+The dataframe should have two columns, one for the name of the feature so we can fetch the feature by name of `str` easily, the other for other metadata of the features. It would be argued that the feature medatadata column contains a hashable, dict-like object. Making it hashable will be convenient for some `pandas` operations. (The hashable implementation is [here](https://github.com/biocore/scikit-bio/pull/1157/files).
+
+
+# Drawbacks
+
+1. The `IntervalIndex` is still in active development and not merged into offical branch.
+There are some methods that don't aren't implemented yet
+(i.e. `IntervalIndex.is_non_overlapping_monotonic`). However, this is in high priority of `pandas`
+milestone and will eventually be merged. We probably set up a dev branch and code based on that.
+This branch can be merged back to master of scikit-bio once `IntervalIndex` gets into pandas.
+
+2. Need to change the API of `positional_metadata`. The proposed data frame with `IntervalIndex` is  different from current implementation, which has feature in column
+and position in row.
+
+3. Need to change the slice implementation. we can update `__getitem__` so that we can fetch the sub sequence by providing an interval or a list of intervals, to make the following code work:
+
+  ```python
+  intervals = seq.positional_metadata.query('feature == gene1').index
+  seq[intervals]
+  ```
+
+  In fact, we recommend to further slim down the operation to have this (even the current slice operation in `scikit-bio` is a little verbose):
+
+  ```python
+  gene1 = seq['gene1']
+  ```
+
+  This is certainly achievable by modifying `__getitem__`.
+
+4. May need to some engineering to combine the per position metadata (such as quality scores) with the interval metadata (such as gene locations). The question is:
+
+   do we really have other per position metadata besides quality score?
```

Comment:
Yes, for example Stockholm format supports arbitrary positional metadata such as posterior probabilities and secondary structure (both `Sequence` and `TabularMSA` have `positional_metadata`).

We propose going with the second option below so that `Sequence` has `metadata`, `positional_metadata`, and `interval_metadata` (that name is not finalized).